File: Classes/C++/CashPayment.cpp
#include "CashPayment.h"
#include <iostream>
#include <iomanip>

CashPayment::CashPayment(float amount) : Payment(amount) {
    change_due = 0.0f;
}

CashPayment::~CashPayment() {}

float CashPayment::getChangeDue() { return change_due; }

float CashPayment::calculateChange(float total) {
    change_due = getAmount() - total;
    if (change_due < 0) {
        cout << "Insufficient funds. Short by $" << fixed << setprecision(2) << (change_due * -1) << endl;
        change_due = 0.0f;
        return 0.0f;
    }
    return change_due;
}

bool CashPayment::processPayment() {
    cout << "Cash payment processed." << endl;
    return true;
}

void CashPayment::displayPaymentInfo() {
    cout << fixed << setprecision(2);
    cout << "Cash Tendered:  $" << getAmount()  << endl;
    cout << "Change Due:     $" << change_due    << endl;
}

File: classes/CashPayment.ts
import { PaymentDef } from "./PaymentDef";

export class CashPayment extends PaymentDef {
  private change_due: number = 0;

  constructor(amount: number) {
    super();
    this.setAmount(amount);
  }

  getAmount(): number        { return this.amount; }
  setAmount(amount: number): void { this.amount = amount; }
  getChangeDue(): number     { return this.change_due; }

  calculateChange(total: number): number {
    this.change_due = this.amount - total;
    if (this.change_due < 0) {
      this.change_due = 0;
      return 0;
    }
    return this.change_due;
  }

  processPayment(): boolean {
    return true;
  }

  displayPaymentInfo(): void {
    console.log(`Cash Tendered: $${this.amount.toFixed(2)}`);
    console.log(`Change Due:    $${this.change_due.toFixed(2)}`);
  }
}
